File: src/main/java/org/snake/models/Vector.java
package org.snake.models;

import java.util.Objects;

public class Vector {
    private final int x;
    private final int y;

    public Vector(int x, int y) {
        this.x = x;
        this.y = y;
    }

    public Vector() {
        this(0, 0);
    }

    public boolean isOrigin() {
        return x == 0 && y == 0;
    }

    public Point add(Point point) {
        return new Point(point.x + this.x, point.y + this.y);
    }

    public Vector add(Vector other) {
        return new Vector(this.x + other.x, this.y + other.y);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Vector vector = (Vector) o;
        return x == vector.x && y == vector.y;
    }

    @Override
    public int hashCode() {
        return Objects.hash(x, y);
    }

    @Override
    public String toString() {
        return String.format("Vector{x=%d, y=%d}", x, y);
    }

}

File: src/main/kotlin/org/snake/models/basic/Vector.kt
package org.snake.models.basic

data class Vector(private val x: Int = 0, private val y: Int = 0) {
    val isOrigin: Boolean = x == 0 && y == 0

    fun add(point: Point): Point = Point(point.x + x, point.y + y)
    fun add(other: Vector): Vector = Vector(x + other.x, y + other.y)

    override fun toString(): String {
        return String.format("Vector{x=%d, y=%d}", x, y)
    }
}